### PR TITLE
cksum: Add untagged option

### LINKED ISSUE #3811: cksum: implement `--untagged`
Used by `tests/misc/sm3sum.pl`

```
--untagged
    create a reversed style checksum, without digest type
```

### PR DESCRIPTION
**Pull request by howjmay**
Closes #3811

### CODE REVIEW

File: src/uu/cksum/src/cksum.rs [hunk 1]
```diff
@@ -106,8 +106,21 @@ fn cksum(fname: &str) -> io::Result<(u32, usize)> {
     }
 }
 
+fn reverse_cksum(checksum: u32) -> u32 {
+    let mut checksum = checksum;
```
Comment (tertsdiepraam) [inline src/uu/cksum/src/cksum.rs:110]: Small nit :)
```suggestion
fn reverse_cksum(mut checksum: u32) -> u32 {
```

**Review by tertsdiepraam — COMMENTED** [review]
Cool, but I'm not sure it's correct. I get the following results with this PR and GNU:
```
❯ cargo run -- cksum --untagged Cargo.lock
1822961466 74712 Cargo.lock
❯ cargo run -- cksum Cargo.lock
1559500086 74712 Cargo.lock
❯ cksum --untagged Cargo.lock
1559500086 74712 Cargo.lock
❯ cksum Cargo.lock
1559500086 74712 Cargo.lock
```

And another issue:
```
❯ cargo run -- cksum --untagged README.md
thread 'main' panicked at 'attempt to subtract with overflow', src/uu/cksum/src/cksum.rs:116:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

As far as I can tell, `--untagged` seems to be a noop if no algorithm is given and we don't support different algorithms yet. Here's GNU when we specify the md5 algorithm:
```
❯ cksum -a md5 README.md
MD5 (README.md) = 08c3964af54e525755b7364b3221a9c8
❯ cksum -a md5 --untagged README.md
08c3964af54e525755b7364b3221a9c8  README.md
```

**Comment by howjmay** [comment]
> Cool, but I'm not sure it's correct. I get the following results with this PR and GNU:
> 
> ```
> ❯ cargo run -- cksum --untagged Cargo.lock
> 1822961466 74712 Cargo.lock
> ❯ cargo run -- cksum Cargo.lock
> 1559500086 74712 Cargo.lock
> ❯ cksum --untagged Cargo.lock
> 1559500086 74712 Cargo.lock
> ❯ cksum Cargo.lock
> 1559500086 74712 Cargo.lock
> ```
> 
> And another issue:
> 
> ```
> ❯ cargo run -- cksum --untagged README.md
> thread 'main' panicked at 'attempt to subtract with overflow', src/uu/cksum/src/cksum.rs:116:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> As far as I can tell, `--untagged` seems to be a noop if no algorithm is given and we don't support different algorithms yet. Here's GNU when we specify the md5 algorithm:
> 
> ```
> ❯ cksum -a md5 README.md
> MD5 (README.md) = 08c3964af54e525755b7364b3221a9c8
> ❯ cksum -a md5 --untagged README.md
> 08c3964af54e525755b7364b3221a9c8  README.md
> ```

Ok it seems that I need to take #3812 together. Let me try to finish #3812 then turn back to this PR?

**Comment by tertsdiepraam** [comment]
Sure! I'll mark this as a draft in the meantime

**Comment by howjmay** [comment]
Ok thanks!

**Comment by sylvestre** [comment]
is it ready for review? thanks

**Comment by howjmay** [comment]
> is it ready for review? thanks

waiting for #4356

**Comment by howjmay** [comment]
done

File: src/uu/cksum/src/cksum.rs (hunk 1 above)
Comment (howjmay) [reply src/uu/cksum/src/cksum.rs]: done

**Comment by howjmay** [comment]
@tertsdiepraam Is there anything I should do for this PR?

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
In my testing GNU has the same output for the default algorithm, but that's not true here.
```
❯ cargo run --no-default-features --features cksum -q -- cksum --untagged Cargo.lock
1214424874 77126 Cargo.lock
❯ cargo run --no-default-features --features cksum -q -- cksum Cargo.lock
1424311826 77126 Cargo.lock
❯ cksum Cargo.lock
1424311826 77126 Cargo.lock
❯ cksum --untagged Cargo.lock
1424311826 77126 Cargo.lock
```

We also definitely need some tests for the other algorithms, which is where `--untagged` should make a big difference in the output.

File: src/uu/cksum/src/cksum.rs [hunk 2]
```diff
@@ -188,21 +202,30 @@ fn digest_read<T: Read>(
...
+}
+
+fn reverse_cksum(checksum: &[u8]) -> Vec<u8> {
+    let sz = checksum.len();
+    let mut ret: Vec<u8> = std::iter::repeat(0).take(sz).collect::<Vec<u8>>();
+    for i in 0..sz {
+        ret[i] = checksum[sz - i - 1].reverse_bits();
     }
+    ret
 }
```
Comment (anastygnome) [inline src/uu/cksum/src/cksum.rs:223]: Hmm, I'd write this without a for loop. 

```rust 
fn reverse_cksum(checksum: &[u8]) -> Vec<u8> {
    checksum.iter().rev().map(|&b| b.reverse_bits()).collect()
}

```

Comment (anastygnome) [reply src/uu/cksum/src/cksum.rs:223]: @tertsdiepraam I forgot whether we try to avoid for loops for uutils, is this a rule here?

Comment (tertsdiepraam) [reply src/uu/cksum/src/cksum.rs:223]: Just use whatever you think is most clear in each situation!

**Comment by howjmay** [comment]
Hi @tertsdiepraam I want to check what `reversed style checksum` means?
The current error I made was caused by I misunderstanding the meaning of `reversed style checksum` or it is just some algorithms that we don't need to reverse them

**Comment by tertsdiepraam** [comment]
I that phrasing is weird. My guess is that they meant only reversing the output, so whether it's filename followed by checksum or checksum followed by filename. This page might help explain: https://www.gnu.org/software/coreutils/manual/html_node/cksum-invocation.html

**Comment by anastygnome** [comment]
AH 
image
that's what it does : 

if the hash function is not a "legacy"  or the default one (bsd,sysv, or crc on my machine) 
it does show it in kind of a math function notation 

`hashfunct(file) = <HASH> `

but if you put untagged it shows it as without the argument 

`<HASH>   file`

```shell 
❯ cksum -a  sha1 a.out
SHA1 (a.out) = 70f26e95e2da7dfb6a223b9adcad311c23781c82
❯ cksum -a  sha1 a.out --untagged
70f26e95e2da7dfb6a223b9adcad311c23781c82  a.out
```

**Comment by tertsdiepraam** [comment]
Could you replace the image with output?

> I don't know if there is a rust way to reuse the code in the sha1sum,... utils, as this is some kind of aliasing : cksum is just doing some maths-like printing it seems.

It does. I was working on this, but it stranded a bit. It's on this branch: https://github.com/uutils/coreutils/compare/main...tertsdiepraam:coreutils:cksum-rewrite

**Comment by anastygnome** [comment]
@tertsdiepraam I removed the image as asked :)

I wanted to do a joke about gnu being bloated, but I will refrain from doing so ;)

**Comment by tertsdiepraam** [comment]
I'm not concerned about breaking changes yet to be honest. We're at version 0.0.17, so I think we can do whatever we want. In my opinion we can just get rid of hashsum and implement the individual utils via `cksum` (though we need feature parity of `cksum` first).

In this case it's not GNU being bloated, it's us :)

File: src/uu/cksum/src/cksum.rs (hunk 2 above)
Comment (sylvestre) [reply src/uu/cksum/src/cksum.rs:223]: @howjmay what do you think here?

**Comment by cakebaker** [comment]
Closing this PR as it has been superseded by #4860. Thanks anyway for your PR.

**Pull request closed without merging**